core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,508 @@
+    }
+
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    BIO *bio = NULL;
+    uint8_t *pem_keypair = NULL;
+    uint32_t key_size = 0;
+
+    uint8_t *rsa_keypair_test = NULL;
+    BIO *bio_test = NULL;
+    RSA *rsa_pubkey_test = NULL;
+
+    pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+    if (pkey_ctx == NULL)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen_init(pkey_ctx) <= 0)
+    {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_RSA_2048:
+        key_size = RSA_2048_KEY_BITS;
+        break;
+    case EH_RSA_3072:
+        key_size = RSA_3072_KEY_BITS;
+        break;
+    case EH_RSA_4096:
+        key_size = RSA_4096_KEY_BITS;
+        break;
+    default:
+        break;
+    }
+
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_ctx, key_size) <= 0)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen(pkey_ctx, &pkey) <= 0)
+    {
+        goto out;
+    }
+
+    bio = BIO_new(BIO_s_mem());
+    if (bio == NULL)
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PUBKEY(bio, pkey))
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PrivateKey(bio, pkey, NULL, NULL, 0, NULL, NULL))
+    {
+        goto out;
+    }
+
+    key_size = BIO_pending(bio);
+    if (key_size <= 0)
+    {
+        goto out;
+    }
+
+    pem_keypair = (uint8_t *)malloc(key_size);
+    if (pem_keypair == NULL)
+    {
+        goto out;
+    }
+
+    if (BIO_read(bio, pem_keypair, key_size) < 0)
+    {
+        goto out;
+    }
+
+    ret = ehsm_create_keyblob(pem_keypair, key_size, (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    if (ret != SGX_SUCCESS)
+    {
+        goto out;
+    }
+out:
+    if (pkey_ctx)
+        EVP_PKEY_CTX_free(pkey_ctx);
+    if (pkey)
+        EVP_PKEY_free(pkey);
+    if (bio)
+        BIO_free(bio);
+
+    memset_s(pem_keypair, key_size, 0, key_size);
+    SAFE_FREE(pem_keypair);
+    return ret;
+}
+
+sgx_status_t ehsm_create_ec_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        return ehsm_calc_keyblob_size(cmk->metadata.keyspec, cmk->keybloblen);
+    }
+
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    BIO *bio = NULL;
+    uint8_t *pem_keypair = NULL;
+    uint32_t key_size = 0;
+
+    pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (pkey_ctx == NULL)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen_init(pkey_ctx) <= 0)
+    {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_EC_P224:
+    case EH_EC_P256:
+    case EH_EC_P384:
+    case EH_EC_P512:
+        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_X9_62_prime256v1) <= 0)
+        {
+            goto out;
+        }
+        break;
+    case EH_SM2:
+        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_sm2) <= 0)
```

Comment:
where do you get the reference?
It's different from the implemetation: https://github.com/intel/intel-sgx-ssl/blob/master/Linux/sgx/test_app/enclave/tests/evp_smx.c#L54

Reply:
function test_EVP_SM2() is my thinking:
https://github.com/openssl/openssl/blob/master/test/evp_extra_test.c

Reply:
change sm2 key generate as https://github.com/intel/intel-sgx-ssl/blob/master/Linux/sgx/test_app/enclave/tests/evp_smx.c#L54